Allow to handle autoclosed characters differently

timfjord opened the pull request:
Adds the `always_treat_brackets_as_autoclosed` setting to control how the autoclosed characters are handled.

The setting is off by default, meaning the behaviour stays the same (following how VSCode handles autoclosed characters).
When set to `true`, the autoclosed characters are always skipped over and auto-removed no matter how they were inserted (following how Sublime Text/Xcode handle this).

https://github.com/zed-industries/zed/assets/471335/304cd04a-59fe-450f-9c65-cc31b781b0db

https://github.com/zed-industries/zed/assets/471335/0f5b09c2-260f-48d4-8528-23f122dee45f

Release Notes:

- Added the setting `always_treat_brackets_as_autoclosed` (default: `false`) to always treat brackets as "auto-closed" brackets, i.e. deleting the pair when deleting start/end, etc. ([#7146](https://github.com/zed-industries/zed/issues/7146)).

timfjord committed:
Allow to handle autoclosed characters differently

timfjord commented on crates/language/src/language_settings.rs:
```diff
@@ -95,6 +95,8 @@ pub struct LanguageSettings {
     /// Which code actions to run on save
     pub code_actions_on_format: HashMap<String, bool>,
+    // Controls how the editor handles the autoclosed characters.
+    pub always_handle_autoclosed_character: bool,
```
I am not 100% sure this is required.
I've followed how the `use_autoclose` is handled but I might have over-engineered it and controlling it globally would be more than enough.

mrnugget replied:
I think it's fine to allow overwriting these per language.

timfjord commented:
@cla-bot check

timfjord committed:
Refactor select_autoclose_pair

timfjord committed:
Fix typos

mrnugget commented:
Hey @timfjord! Sorry for the late response here, but thanks for opening the PR. I watched both videos, but I'm still a bit unsure what exactly is the difference in the behavior, since I can't see your keystrokes. I'll take a closer look at the code tomorrow.

timfjord commented:
Hey @mrnugget 

Thanks for coming back to me.
I've re-uploaded both videos with a keystrokes visualiser enabled.

In a nutshell, when the flag is set to `false`(default behaviour) skipping over and auto-removing the closing characters happens only for the auto-inserted characters that were literally just entered.
So if you move your cursor you won't be able to skip over and auto-remove them(see the `when-false` video in the description, where I tried to reproduce this by moving my cursor after entering the `{` character).
This is how VSCode handles this.
The easiest way to reproduce this is to select the `Rust` language in the language selection(`⌘K  M`), enter `{` twice, then move the cursor(say move right and then left) and then press `}`. The editor won't let you skip over, instead the `}` will be entered.
The same goes for the auto-removal. After moving the cursor, the editor doesn't allow to auto-remove the pairs.

Other editors like Sublime Text, Xcode and probably others follow a bit different, I would say, more pragmatic approach. 
They allow you to skip over and auto-remove at any time, no matter when the auto-inserted characters were entered (even if you open a new file and want to delete some `{}` pairs). 
The idea of this PR is to allow to enable this mode for Zed.
I tried to illustrate this in the `when-true` video (this is how the editor performs when the flag is set to `true`) where the closing characters are skipped and auto-removed over after the cursor moves.

mrnugget commented on crates/editor/src/editor_tests.rs:
```diff
@@ -4519,6 +4519,120 @@ async fn test_autoclose_pairs(cx: &mut gpui::TestAppContext) {
+                        close: true,
+                        newline: true,
+                    },
+                    BracketPair {
+                        start: "[".to_string(),
+                        end: "]".to_string(),
+                        close: false,
+                        newline: true,
+                    },
+                    BracketPair {
+                        start: "\"".to_string(),
+                        end: "\"".to_string(),
+                        close: true,
+                        newline: false,
+                    },
```
Do you need these here? Seems like you don't use them in the test.

timfjord replied:
Probably we don't, copied from the the above

mrnugget commented on crates/editor/src/editor_tests.rs:
```diff
@@ -5121,6 +5235,121 @@ async fn test_delete_autoclose_pair(cx: &mut gpui::TestAppContext) {
+                        close: true,
+                        newline: true,
+                    },
+                    BracketPair {
+                        start: "/*".to_string(),
+                        end: " */".to_string(),
+                        close: true,
+                        newline: true,
+                    },
```
Same as above: if we don't use it, let's not add it here in the tests.

mrnugget commented on assets/settings/default.json:
```diff
@@ -89,6 +89,12 @@
+  // happen only for auto-inserted characters.
+  // Otherwise(when `true`), the closing characters are always skipped over and auto-removed
+  // no matter how they were inserted.
+  "always_handle_autoclosed_character": false,
```
I know it's not the best name either, but `treat_all_brackets_as_auto_closed` would explain things better for me. What do you think? Or `always_treat_brackets_as_autoclosed`?

mrnugget replied:
I think `brackets` is also the word we use in other language configurations, even though not all of them are brackets per se. So for consistency's case, I think it makes sense to use `brackets` :)

timfjord replied:
Great, I will rename the flag to `always_treat_brackets_as_autoclosed` then

mrnugget reviewed (approved):
I now get it, thanks for your explanations! I left some comments. I think you can clean up the tests a bit.

Also: can you make sure that the `Release notes` in your PR are updated? Ideally it explains what has changed and why and how users can enable it (it'll show up in the changelog)

timfjord commented:
@mrnugget Thanks for reviewing the PR.

I have one general question. I was testing this PR and noticed that when I was changing the flag in the tested editor(the one I get after running `cargo run`) the editor didn't react.
For example, when I did the same with the `use_autoclose` flag it reacted immediately.
So now I wonder if maybe I need to somehow let the editor know about the new flag.

So far, I've added the flag to the `Editor`, `LanguageSettings`, and `LanguageSettingsContent` structs and it is working fine in tests(partly because I set the flag directly).
The config gets reloaded each time it is updated, so maybe I need to add something somewhere

Just in case, I couldn't find anything like that for the `use_autoclose` flag.

mrnugget commented:
@timfjord I'm not 100% sure myself, but I do think that some of the settings (such as `use_autoclose`) are evaluated each time in the editor and the value on `editor` itself is only cached.

But look at `self.show_breadcrumbs` here and how it gets updated: https://github.com/zed-industries/zed/blob/f1c3335a9090cdfa69ab6b84c3ad42de7a79e9b5/crates/editor/src/editor.rs#L9282

That's in a callback when the settings update, so you probably want to change your value here too.

timfjord commented:
Thanks for the heads-up
I think I saw this one, but since there was no `use_autoclose` (the one that gets reloaded) I didn’t debug it. Anyway, will spend more time debugging this

timfjord committed:
Rename the flag to `always_treat_brackets_as_autoclosed`

timfjord committed:
Clean up tests

timfjord commented:
@mrnugget I've addressed your feedback:
- renamed the flag to `always_treat_brackets_as_autoclosed`
- simplified the tests
- and also resolved the conflicts

mrnugget commented:
@timfjord I again didn't get to review it today, sorry! I'll try to get to it tomorrow!

timfjord commented:
No worries, thanks for the update.

timfjord commented:
> I tested this locally, but I can't get it to work. `self.always_treat_brackets_as_autoclosed` in `Editor` is always false for me, no matter what I set in settings. Did you try this manually? Does it only work when you have language override?
> 
> I think it should change when you update the settings, which is what I meant here: [#8666 (comment)](https://github.com/zed-industries/zed/pull/8666#issuecomment-1986212803)

That was the case for me too, I was going to debug it, but probably forgot about that.
Let me work on that and I will let you know when it is ready

timfjord committed:
Add docs

timfjord committed:
Change position in the LanguageSettings

timfjord commented:
@mrnugget Sorry for bombarding you with the messages, but I think I've figured out why it doesn't work.
It is a combination of different factors, and the most important being that we use `&&` here:
```rust
let always_treat_brackets_as_autoclosed = self
    .always_treat_brackets_as_autoclosed
    && snapshot
        .settings_at(selection.start, cx)
        .always_treat_brackets_as_autoclosed;
```
and since the default value is `false`, it always returns `false`, no matter what is the value in the language settings
(it works for `use_autoclose`, because the default value for `use_autoclose` is `true`, which, by the way, might lead to a bug, but that's a different story).
So I will continue my research and notify you with the results.

mrnugget commented:
> Sorry for bombarding you with the messages

Don't worry about my notifications :)

timfjord committed:
Remove always_treat_brackets_as_autoclosed from editor.rs

timfjord commented:
@mrnugget It is now working, please take a look, when you have a moment.

As I thought, the problem was with that `set_always_treat_brackets_as_autoclosed` setter in the `Editor`.
I copied that logic from `use_autoclose` but the use case was a bit different, because `set_use_autoclose` is used in some other parts of the editor and this is how the `use_autoclose` flag is forced to be always `false`(for example in the `MessageEditor`).
The setter wasn't needed in my case anyway(as well as the getter), plus the chosen default value(`false`) caused the flag to be always `false`, that's why it worked in tests(where I could explicitly call the setter and set it to `true`) and didn't work in the real editor, where all values were taken from the settings.

mrnugget committed:
Remove nested conditions and update docs a bit

mrnugget commented:
Thanks @timfjord! I hopped on the branch and removed two nested conditionals (that were already in the original code, but jumped out to me). Also added an example to the docs. Will merge once CI is through.

timfjord commented:
Great, thanks for the update 👍🏻
